File: bridge/WaveShaperNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createWaveShaperNode(AudioContext* context) {
    auto node = std::make_shared<WaveShaperNode>(*context);
    return keepNode(node);
}

DART_EXPORT void WaveShaperNode_setCurve(const int nodeId, const float* curve) {
    auto node = std::static_pointer_cast<WaveShaperNode>(getNode(nodeId));
    if(node) {
        std::vector<float> vecCurve(curve, curve + sizeof(curve)/sizeof(float));
        node->setCurve(vecCurve);
    }
}

File: lib/src/lab_sound/wave_shaper_node.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'audio_context.dart';
import 'audio_node.dart';
import 'lab_sound.dart';

class WaveShaperNode extends AudioNode {
  WaveShaperNode(AudioContext ctx): super(ctx, LabSound().createWaveShaperNode(ctx.pointer));

  setCurve(List<double> curve) {
    final curvePtr = malloc.allocate<Float>(curve.length);
    for (int i = 0; i < curve.length; i++) {
      curvePtr[i] = curve[i];
    }
    LabSound().WaveShaperNode_setCurve(nodeId, curvePtr);
    malloc.free(curvePtr);
  }
}